scripts/cpp/ompl_integration.cc

```diff
@@ -0,0 +1,242 @@
+#include <vector>
+#include <array>
+#include <utility>
+#include <iostream>
+
+#include <vamp/collision/factory.hh>
+#include <vamp/planning/validate.hh>
+
+#include <vamp/robots/panda.hh>
+
+#include <ompl/base/MotionValidator.h>
+#include <ompl/base/ProblemDefinition.h>
+#include <ompl/base/SpaceInformation.h>
+#include <ompl/base/StateValidityChecker.h>
+#include <ompl/base/spaces/RealVectorStateSpace.h>
+#include <ompl/base/objectives/PathLengthOptimizationObjective.h>
+#include <ompl/geometric/PathSimplifier.h>
+#include <ompl/geometric/planners/informedtrees/BITstar.h>
+#include <ompl/util/Exception.h>
+
+namespace ob = ompl::base;
+namespace og = ompl::geometric;
+
+using Robot = vamp::robots::Panda;
+static constexpr std::size_t dimension = Robot::dimension;
+using Configuration = Robot::Configuration;
+static constexpr const std::size_t rake = vamp::FloatVectorWidth;
+using EnvironmentInput = vamp::collision::Environment<float>;
+using EnvironmentVector = vamp::collision::Environment<vamp::FloatVector<rake>>;
+
+static const std::array<float, dimension> zeros = {0., 0., 0., 0., 0., 0., 0.};
+static const std::array<float, dimension> ones = {1., 1., 1., 1., 1., 1., 1.};
+
+// Start and goal configurations
+static const std::array<float, dimension> start = {0., -0.785, 0., -2.356, 0., 1.571, 0.785};
+static const std::array<float, dimension> goal = {2.35, 1., 0., -0.8, 0, 2.5, 0.785};
+
+// Spheres for the cage problem
+static const std::vector<std::array<float, 3>> problem = {
+    {0.55, 0, 0.25},
+    {0.35, 0.35, 0.25},
+    {0, 0.55, 0.25},
+    {-0.55, 0, 0.25},
+    {-0.35, -0.35, 0.25},
+    {0, -0.55, 0.25},
+    {0.35, -0.35, 0.25},
+    {0.35, 0.35, 0.8},
+    {0, 0.55, 0.8},
+    {-0.35, 0.35, 0.8},
+    {-0.55, 0, 0.8},
+    {-0.35, -0.35, 0.8},
+    {0, -0.55, 0.8},
+    {0.35, -0.35, 0.8},
+};
+
+// Radius for obstacle spheres
+static constexpr const float radius = 0.2;
+
+// Convert an OMPL state into a VAMP vector
+inline static auto ompl_to_vamp(const ob::State *state) -> Configuration
+{
+    // Create an aligned memory buffer to load VAMP vector from
+    alignas(Configuration::S::Alignment)
+        std::array<typename Configuration::S::ScalarT, Configuration::num_scalars>
```

Comment:
Maybe make static? Doesn't matter too much; we're not trying to maximize perf here.

Reply:
Some OMPL operations may potentially be multi-threaded, not worth the headache of potential bug reports later on.